meta/include/gnuradio-4.0/meta/UncertainValue.hpp

```diff
@@ -0,0 +1,331 @@
+
+    // Default constructor
+    constexpr UncertainValue() noexcept = default;
+
+    constexpr UncertainValue(T value_, T uncertainty_) noexcept : value(value_), uncertainty(uncertainty_) {}
+
+    explicit(false) constexpr UncertainValue(T value_) noexcept : value(value_), uncertainty(static_cast<T>(0)) {}
+
+    constexpr UncertainValue(const UncertainValue &) noexcept = default;
+    constexpr UncertainValue(UncertainValue &&) noexcept      = default;
+    constexpr UncertainValue &
+    operator=(const UncertainValue &) noexcept
+            = default;
+    constexpr UncertainValue &
+    operator=(UncertainValue &&) noexcept
+            = default;
+    ~UncertainValue() = default;
+
+    constexpr UncertainValue &
+    operator=(const T &other) noexcept {
+        value       = other;
+        uncertainty = static_cast<T>(0);
+        return *this;
+    }
+
+    constexpr UncertainValue &
+    operator=(T &&other) noexcept {
+        value       = std::move(other);
+        uncertainty = static_cast<T>(0);
+        return *this;
+    }
+
+    auto operator<=>(UncertainValue const&) const = default;
+};
+
+template<typename T>
+UncertainValue(T, T) -> UncertainValue<T>;
+
+template<typename T>
+concept UncertainValueLike = requires(T t) {
+    typename T::value_type;                                           // UncertainValue must have a nested value_type
+    { t.uncertainty } -> std::convertible_to<typename T::value_type>; // Must have an 'uncertainty' field
+};
+
+namespace detail {
+template<typename T>
+struct UncertainValueTypeType {
+    using type = T;
+};
+
+template<typename T>
+struct UncertainValueTypeType<UncertainValue<T>> {
+    using type = T;
+};
+} // namespace detail
+
+template<typename T>
+using UncertainValueType_t = detail::UncertainValueTypeType<T>::type;
+
+/********************** some basic math operation definitions *********************************/
+
+template<typename T, typename U>
+    requires UncertainValueLike<T> || UncertainValueLike<U>
+[[nodiscard]] inline constexpr auto
+operator+(const T &lhs, const U &rhs) noexcept {
+    using ValueTypeT = UncertainValueType_t<T>;
+    using ValueTypeU = UncertainValueType_t<U>;
+    if constexpr (UncertainValueLike<T> && UncertainValueLike<U>) {
+        using ResultType = decltype(lhs.value + rhs.value);
+        if constexpr (complex_like<ValueTypeT> || complex_like<ValueTypeU>) {
+            // we are dealing with complex numbers -> use the standard uncorrelated calculation.
+            ResultType newUncertainty = { std::hypot(std::real(lhs.uncertainty), std::real(rhs.uncertainty)), std::hypot(std::imag(lhs.uncertainty), std::imag(rhs.uncertainty)) };
+            return UncertainValue<ResultType>{ lhs.value + rhs.value, newUncertainty };
+        } else {
+            // both ValueType[T,U] are arithmetic uncertainties
+            return UncertainValue<ResultType>{ lhs.value + rhs.value, std::hypot(lhs.uncertainty, rhs.uncertainty) };
```

Comment:
IMO, this is not universal. The numeric error can be calculated in different ways, not only with standard deviation.

Reply:
we handle only uncorrelated standard deviations (i.e. no asymmetric or correlated values).